Inserting Test Data without File Read and Write

kramosss opened the pull request:
# Description

Avoids reading and writing to a file for test data and instead directly inserts the test data into the data base. 

Partly Addresses #141 and #1131 

## Type of change

(Check the ones that apply by placing an "x" instead of the space in the [ ] so it becomes [x])

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

(Note what you have done by placing an "x" instead of the space in the [ ] so it becomes [x]. It is hoped you do all of them.)

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

## Limitations

The test data still has some issue running in parallel and will need to be worked on in the future.

kramosss committed:
update to LG4 Test

kramosss committed:
Merge branch 'development' of https://github.com/kramosss/OED into development

kramosss committed:
update from origin

kramosss committed:
Merge branch 'development' of https://github.com/kramosss/OED into development

kramosss committed:
update with origin

kramosss committed:
Merge branch 'development' of https://github.com/kramosss/OED into development

kramosss committed:
Merge branch 'development' of https://github.com/kramosss/OED into development

kramosss committed:
Merge branch 'OpenEnergyDashboard:development' into development

kramosss committed:
Merge branch 'OpenEnergyDashboard:development' into development

kramosss committed:
Changes made to insertMeters, generateSineData, and  generateFourDayTestingData. LoadGeneratedInput file added.

huss committed:
update for some issues

- fix brace issue.
- fix function import.
- Misc format

kramosss committed:
Moved the meter data for the generate functions

kramosss committed:
Took out comments and formatted the code, while
also creating a function for test generation and
inserting the meter data

kramosss committed:
Took out one of the comments for formatting

kramosss committed:
Merge branch 'OpenEnergyDashboard:development' into testGenNoFile

kramosss committed:
Merge branch 'OpenEnergyDashboard:development' into testGenNoFile

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -13,6 +13,7 @@ const { getConnection } = require('../db');
 const { redoCik } = require('../services/graph/redoCik');
 const { refreshAllReadingViews } = require('../services/refreshAllReadingViews');
 const fs = require('fs').promises;
+const conn = getConnection();
```
In every case I checked, OED uses ``getConnection()`` within a function and not outside. I think this is better practice so this should be added where needed within functions.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -41,11 +60,12 @@ async function generateSineTestingData(frequency = 15, amplitude = 1) {
+	// Store data in a variable
+	const generatedData = await generateSine(startDate, endDate, options);
+	// Return the generatedData in the function
+	return generatedData;
```
Pease add an @return in the JSDoc for this.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -41,11 +60,12 @@ async function generateSineTestingData(frequency = 15, amplitude = 1) {
+	return generatedData;
 }
 
 /**
```
Update this comment so say it puts the values into the database rather than saves to file.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -106,13 +141,28 @@ async function generateSineSquaredTestingData(amplitude = 1) {
-		filename: path.join(__dirname, '../test/db/data/automatedTests/' + frequency.toString() + 'FreqCosineTestData.csv')
 	};
-	await generateCosine(startDate, endDate, options);
+	insertData(startDate, endDate, options, meterData, conn)
```
I think this generates sine instead of cosine data.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -90,10 +126,9 @@ async function generateSineSquaredTestingData(amplitude = 1) {
+	insertData(startDate, endDate, options, meterData, conn)
 }
 
 /**
```
See above about update for putting into DB.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -106,13 +141,28 @@ async function generateSineSquaredTestingData(amplitude = 1) {
+	insertData(startDate, endDate, options, meterData, conn)
 }
 
 /**
```
Into DB.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -157,10 +223,9 @@ async function generateCosineSquaredTestingData(amplitude = 1) {
 		squared: true // Option set to true because want cosine *squared* data.
 	};
-	await generateCosine(startDate, endDate, options);
+	insertData(startDate, endDate, options, meterData, conn);
```
I think this generates sine instead of cosine data.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -179,14 +244,29 @@ async function generateCosineSquaredTestingData(amplitude = 1) {
 async function generateFourDayTestingData() {
```
Update comment above about into DB. This needs to happen for a lot of functions below.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -234,14 +346,29 @@ async function generateTwentyThreeMinuteTestingData() {
+	insertData(startDate, endDate, options, meterData, conn);
 }
 
 /**
```
While not part of the standard, generated data, it can be called via npm. The generated file is very large so I think we should see if it could be stored directly into the DB as with other functions (hopefully the variable is not too large).

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -273,24 +400,24 @@ async function generateTestingData() {
 	await generateFourHourTestingData();
 
-	// Generates 1 year of sinusoidal data with data points at 23-minute intervals
+	// // Generates 1 year of sinusoidal data with data points at 23-minute intervals
```
This and following comments have double // that should be removed.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -273,24 +400,24 @@ async function generateTestingData() {
+	// // Generates 2 years of *squared* cosinusoidal data with an amplitude of 2.5 and with data points at 1-day intervals.
 	await generateCosineSquaredTestingData(2.5);
 }
 
```
Should say goes into meters and not file.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -166,10 +166,29 @@ function generateSineData(startTimeStamp, endTimeStamp, options = {}) {
+
+	return generatedData;
+
+	//return (_.zip(sineValues, startDates, endDates));
```
The commented out code should probably be deleted.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -166,10 +166,29 @@ function generateSineData(startTimeStamp, endTimeStamp, options = {}) {
+	return generatedData;
+
+	//return (_.zip(sineValues, startDates, endDates));
+} 
```
Remove extra space at end of line.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -255,9 +274,14 @@ async function generateSine(startTimeStamp, endTimeStamp, options = {}) {
+		return sineData; 
 	} catch (error) {
-		log.error(`Failed to generate sine data for file: ${chosenOptions.filename}.`, error);
+		//log.error(`Failed to generate sine data for file: ${chosenOptions.filename}.`, error);
```
Remove commented out line.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -255,9 +274,14 @@ async function generateSine(startTimeStamp, endTimeStamp, options = {}) {
 			chosenOptions.maxAmplitude = chosenOptions.maxAmplitude * scale;
 		}
-		await writeToCsv(generateSineData(startTimeStamp, endTimeStamp, chosenOptions), chosenOptions.filename);
+		//await writeToCsv(generateSineData(startTimeStamp, endTimeStamp, chosenOptions), chosenOptions.filename);
```
Remove commented out line.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -229,7 +248,7 @@ async function generateSine(startTimeStamp, endTimeStamp, options = {}) {
 		maxAmplitude: options.maxAmplitude || 2,
 		noShift: options.noShift || false,
-		filename: options.filename || 'test.csv',
+		//filename: options.filename || 'test.csv',
```
Remove commented out line.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -255,9 +274,14 @@ async function generateSine(startTimeStamp, endTimeStamp, options = {}) {
+		//Generate the data
+		const sineData = generateSineData(startTimeStamp, endTimeStamp, chosenOptions);
+		//return data in the function
+		return sineData; 
```
Add description as @return in JSDoc above and include error condition below.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -284,7 +308,7 @@ async function generateCosine(startTimeStamp, endTimeStamp, options = {}) {
 		periodLength: options.periodLength || { day: 1 },
 		maxAmplitude: options.maxAmplitude || 2,
-		noShift: options.noShift || false,
+		noShift: options.noShift || false, 
```
Remove space at end of line.

huss commented on src/server/data/generateTestingData.js:
```diff
@@ -311,9 +335,14 @@ async function generateCosine(startTimeStamp, endTimeStamp, options = {}) {
 			chosenOptions.maxAmplitude = chosenOptions.maxAmplitude * scale;
 		}
-		await writeToCsv(generateSineData(startTimeStamp, endTimeStamp, chosenOptions), chosenOptions.filename);
+		//await writeToCsv(generateSineData(startTimeStamp, endTimeStamp, chosenOptions), chosenOptions.filename);
```
See sine function above and make similar changes here.

huss commented on src/server/services/pipeline-in-progress/loadGeneratedInput.js:
```diff
@@ -0,0 +1,63 @@
+	readingGap,
+	readingLengthVariation,
+	isEndOnly,
+	headerRow,
```
headerRow is not a parameter that is used. Should it be removed here and JSDoc? Calls to function would need to be updated.

huss commented on src/server/services/pipeline-in-progress/loadGeneratedInput.js:
```diff
@@ -0,0 +1,63 @@
+const { log } = require('../../log');
+
+/**
+ * Read a CSV file and select needed column to return an array of reading value and reading time
```
I don't think this function works with CSV.

huss commented on src/server/util/insertData.js:
```diff
@@ -297,6 +298,7 @@ async function insertStandardConversions(conn) {
  * Note the values provided for the keys are not checked for validity.
  * @param {[{}]} metersToInsert key:value pairs of meter values in array with entry for each meter
  * @param {*} conn database connection to use
+ * @param {string} dataSource - indicates the source of the data, whether its from a file or variable 
```
This is not a parameter.

huss commented on src/server/util/insertData.js:
```diff
@@ -310,14 +312,24 @@ async function insertMeters(metersToInsert, conn) {
 
 		// Check that needed keys are there.
-		const requiredKeys = ['name', 'file'];
+		//const requiredKeys = ['name', 'file'];
```
Remove commented out line.

huss commented on src/server/util/insertData.js:
```diff
@@ -410,36 +432,64 @@ async function insertMeters(metersToInsert, conn) {
+					// Create loadGeneratedInput and call it here   
+					// Pass a new variable like meter.data instead of file name, data that needs to map, function uses that data 
+					// Otherwise if there is no file, load data from variable
+					await loadGeneratedInput(
```
The arguments do not match all the parameters in the actual function.

huss reviewed (changes requested):
Thanks to @kramosss for their first OED contribution. I've made several comments to consider. Note I still need to look into the await issue/failing test cases but wanted to get these comments out. I will also test carefully once everything is in place. Please let me know if you need anything.

huss committed:
fix two generate tests

Change test to use keys instead of array values for new code.

huss committed:
remove write to file tests

The test data is placed directly into DB and not file so these
tests are obsolete.

huss committed:
fix generate tests

Test directly test data instead of from file per new code.

huss committed:
fix unit test for generation

Modifies so uses new generation format.

huss committed:
fix up async test data

- The generate functions are not async.
- Minor formatting.
- await on inserting meters and insertData.
- Added info on removing all DB values to rerun testData.

huss committed:
Make data insertion mostly async

Note the way to wait for meter insertion has changed.

huss committed:
wrap meter insert into Promise.all

huss committed:
fix cosine generation

huss committed:
1 min now directly into DB without file

Also added msg that generating data for meter.

huss committed:
remove writeToCsv() since no longer used

huss committed:
remove deleteFile +

- deleteFile not needed when no file written so removed.
- Added generation msg for amp meters that was missed before.

huss commented:
I made changes to deal with async code, get test cases to work and other items. I tried to update all PR comments so they are in sync with these changes and resolved as appropriate. Some comments are marked outdated due to shifting line so @kramosss let me know if it is unclear what they apply to.

kramosss committed:
Merge branch 'OpenEnergyDashboard:development' into testGenNoFile

kramosss committed:
Updated code should have all the changes.

kramosss commented:
I updated the code and it should have all the changes, feel free to let me know if I missed something or if there are any other changes to be made.

huss committed:
update comments

huss commented:
Thanks to @kramosss for the quick update. Everything looks good. Congratulations on another accepted contribution to OED.
